test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,516 @@
+    local term = g.server_2:synchro_queue_term()
+    promote(g.server_2)
+    g.server_1:wait_synchro_queue_term(term + 1)
+
+    local ok, err = fiber_join(g.server_1, f)
+    luatest.assert(not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'Interfering promote fails')
+
+    wait_sync({g.server_1, g.server_2, server_3})
+    box_cfg_update(g.cluster.servers, g.box_cfg)
+    server_3:drop()
+    promote(g.server_1)
+    g.server_1:exec(function() box.space.test:drop() end)
+    demote(g.server_1)
+    wait_sync(g.cluster.servers)
+end
+
+-- Demoting should fail if it is interrupted from another server
+-- while waiting for synchro queue being emptied.
+g_common.test_limbo_full_interfering_demote = function(g)
+    promote(g.server_1)
+    wait_sync(g.cluster.servers)
+
+    g.server_1:exec(function()
+        box.schema.create_space('test', {is_sync = true}):create_index('pk')
+    end)
+
+    box_cfg_update({g.server_1}, {
+        replication_synchro_quorum = 3,
+        replication_synchro_timeout = 1000,
+    })
+
+    box_cfg_update({g.server_2}, {
+        replication_synchro_timeout = 0.1,
+    })
+
+    g.server_1:exec(function()
+        local s = box.space.test
+        require('fiber').create(s.replace, s, {1})
+    end)
+    wait_sync(g.cluster.servers)
+
+    -- Start demoting server_1 and interrupt it from server_2
+    local f = demote_start(g.server_1)
+    local term = g.server_1:synchro_queue_term()
+    g.server_2:exec(function() pcall(box.ctl.promote) end)
+    g.server_1:wait_synchro_queue_term(term + 1)
+
+    local ok, err = fiber_join(g.server_1, f)
+    luatest.assert(not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'Interfering demote fails')
+
+    wait_sync(g.cluster.servers)
+    promote(g.server_1)
+    g.server_1:exec(function() box.space.test:drop() end)
+    demote(g.server_1)
+    wait_sync(g.cluster.servers)
+end
+
+-- Promoting should fail if synchro queue replication timeouts during it
+g_common.test_fail_limbo_ack_promote = function(g)
+    box_cfg_update({g.server_1}, {
+        replication_synchro_quorum = 3,
+    })
+
+    box_cfg_update({g.server_2}, {
+        replication_synchro_quorum = 3,
+        replication_synchro_timeout = 1000,
+    })
+
+    -- fill synchro queue on server_1
+    promote(g.server_2)
+    g.server_2:exec(function()
+        local s = box.schema.create_space('test', {is_sync = true})
+        s:create_index('pk')
+        require('fiber').create(s.replace, s, {1})
+    end)
+    wait_sync(g.cluster.servers)
+
+    -- start promoting with default replication_synchro_timeout,
+    -- wait until promote reaches waiting for limbo_acked,
+    -- make it timeout by lowering replication_synchro_timeout
+    local ok, err = g.server_1:exec(function()
+        local fiber = require('fiber')
+        local f = fiber.new(function() box.ctl.promote() end)
+        f:set_joinable(true)
+        box.cfg{replication_synchro_timeout = 0.01}
```

Comment:
Updating the synchro timeout won't yield. Which means this is the same as doing:
```Lua
box.cfg{replication_synchro_timeout = 0.01}
box.ctl.promote()
```
without any fibers. Are you sure this is what you intended to do?

Reply:
After term has been written to WAL promote() waits until limbo is acked for 2*replication_synchro_timeout, during this wait i I change replication_synchro_timeout to lower value, which makes waiting transactions to timeout, leading to promote failure.

Reply:
You are missing the point - I meant that you didn't need `fiber.new()` in the old code. It won't start the fiber immediately. Hence in your old code you didn't change the timeout *during* promote. You changed it *before* the promote. Because the fiber created with `fiber.new()` will actually start only after a yield, which you didn't have before the `box.cfg` below.